File: ServerSide/BallPassSchedulePattern/Players_QrcodeTypes/s_QRCodeMaskApplicatorPlayerDir/QRCodeMaskApplicatorPlayer.py
import os, sys
sys.path.append("../..")
from Players_CommonPlayers.SuperPlayerDir.SuperPlayer import SuperPlayer

class QRCodeMaskApplicatorPlayer(SuperPlayer):  # 新しい名前！
    def __init__(self):
        super().__init__()
        self.my_name = None  # 初期化は必ずNone
        self.qr_map_2Dlist = None  # QRコードの2Dリストを保持

    def return_my_name(self):
        return "QRCodeMaskApplicatorPlayer"  # 新しい名前！

    def apply_mask(self, matrix, mask_bit, i, j):
        """
        2次元リスト matrix に対して、指定された mask_bit に基づいて (i, j) のビットを反転させる。
        """
        if mask_bit == "000":
            if (i + j) % 2 == 0:
                matrix[i][j] = 1 - matrix[i][j]  # ビットを反転
        elif mask_bit == "001":
            if i % 2 == 0:
                matrix[i][j] = 1 - matrix[i][j]
        elif mask_bit == "010":
            if j % 3 == 0:
                matrix[i][j] = 1 - matrix[i][j]
        elif mask_bit == "011":
            if (i + j) % 3 == 0:
                matrix[i][j] = 1 - matrix[i][j]
        elif mask_bit == "100":
            if ((i // 2) + (j // 3)) % 2 == 0:
                matrix[i][j] = 1 - matrix[i][j]
        elif mask_bit == "101":
            if (i * j) % 2 + (i * j) % 3 == 0:
                matrix[i][j] = 1 - matrix[i][j]
        elif mask_bit == "110":
            if ((i * j) % 2 + (i * j) % 3) % 2 == 0:
                matrix[i][j] = 1 - matrix[i][j]
        elif mask_bit == "111":
            if ((i * j) % 3 + (i + j) % 2) % 2 == 0:
                matrix[i][j] = 1 - matrix[i][j]
        return matrix

    def main(self, matrix, mask_bit):
        """
        2次元リスト matrix (25x25) に対して、mask_bitに基づき全座標のビットを反転させて変換後のリストを返す。
        """
        for i in range(len(matrix)):   # 行の数
            for j in range(len(matrix[i])):  # 列の数
                matrix = self.apply_mask(matrix, mask_bit, i, j)  # 各座標ごとにマスク適用

        self.qr_map_2Dlist = matrix
        self.one_time_world_instance.qrCodeMaskApplicatorPlayer = self  # 結果を保持
        return matrix

File: InstanceSide/BallPassSchedulePattern/Players_QRcodeTypes/s_QRCodeMaskApplicatorPlayerDir/QRCodeMaskApplicatorPlayer.cs
using UdonSharp;
using UnityEngine;

public class QRCodeMaskApplicatorPlayer : SuperPlayer
{
    [SerializeField] private int[][] qrMap2DList;  // QRコードの2Dリスト
    [SerializeField] private int[][] updatedQrMap2DList;  // 更新済みのQRコードマップ
    [SerializeField] private RinaNumpy rinaNumpy;  // RinaNumpyをアタッチ

    public override string ReturnMyName()
    {
        return "QRCodeMaskApplicatorPlayer";  // クラス名を返す
    }

    public int[][] ApplyMask(int[][] matrix, string maskBit, int i, int j)
    {
        // 指定された maskBit に基づき (i, j) のビットを反転する
        if (rinaNumpy == null)
        {
            Debug.LogError("RinaNumpy is not assigned.");
            return matrix;  // rinaNumpyが設定されていない場合は変更しない
        }

        float[] maskValues = rinaNumpy.ConvertToFloatArrayFromBitString(maskBit);

        // 各mask_bit条件を評価し、ビットを反転
        if (maskValues[0] == 0 && (i + j) % 2 == 0)
        {
            matrix[i][j] = 1 - matrix[i][j];
        }
        else if (maskValues[1] == 0 && i % 2 == 0)
        {
            matrix[i][j] = 1 - matrix[i][j];
        }
        else if (maskValues[2] == 0 && j % 3 == 0)
        {
            matrix[i][j] = 1 - matrix[i][j];
        }
        // 他の条件もrinaNumpyと統合可能
        return matrix;
    }

    public override void ExecuteMain()
    {
        if (rinaNumpy == null)
        {
            Debug.LogError("RinaNumpy is not assigned. Please attach it in the inspector.");
            return;
        }

        // Unityエディタで設定された2Dリストを取得
        int[][] matrix = updatedQrMap2DList;

        // マスクビットを「000」に固定
        string maskBit = "000";

        // マトリックスにマスクを適用
        for (int i = 0; i < matrix.Length; i++)
        {
            for (int j = 0; j < matrix[i].Length; j++)
            {
                matrix = ApplyMask(matrix, maskBit, i, j);
            }
        }

        // RinaNumpyで行列を操作する例
        float meanValue = rinaNumpy.Mean_FloatArray(rinaNumpy.ConvertToFloatArray(matrix));

        Debug.Log($"Mean value of QR matrix: {meanValue}");

        // 結果を更新
        qrMap2DList = matrix;
    }
}
